New track calo skimmer module for ICARUS calibration stuff.

gputnam opened the pull request:
New module. Orthogonal to existing stuff. Depends on https://github.com/SBNSoftware/sbnobj/pull/17.

wesketchum commented on sbncode/Calibration/TrackCaloSkimmerSelectStoppingTrack_tool.cc:
```diff
@@ -0,0 +1,183 @@
+  // compute the median dqdx of the last 5 cm
+  std::vector<double> endp_dqdx;
+  for (const sbn::HitInfo &h: t.hits2) {
+    if (h.oncalo && h.rr < 5.) endp_dqdx.push_back(h.dqdx);
```
Suggestion: make configurable?

gputnam replied:
Fixed.

wesketchum commented on sbncode/Calibration/TrackCaloSkimmerSelectStoppingTrack_tool.cc:
```diff
@@ -0,0 +1,183 @@
+    }
+  }
+
+  std::cout << "Track end: " << t.end_x << " " << t.end_y << " " << t.end_z << " is FID: " << end_is_fid << std::endl;
```
Wrap in a verbosity option.

gputnam replied:
Fixed.

wesketchum commented on sbncode/Calibration/TrackCaloSkimmerSelectStoppingTrack_tool.cc:
```diff
@@ -0,0 +1,183 @@
+    (t.hit_min_time_p2_tpcW < 0. || t.hit_min_time_p2_tpcW > fFidTickMin) &&
+    (t.hit_max_time_p2_tpcW < 0. || t.hit_max_time_p2_tpcW < fFidTickMax);
+
+  std::cout << "Track times: " << t.hit_min_time_p2_tpcE << " " << t.hit_max_time_p2_tpcE << " " << t.hit_min_time_p2_tpcW << " " << t.hit_max_time_p2_tpcW << " is FID: " << time_is_fid << std::endl;
```
Wrap in a verbosity option.

gputnam replied:
Fixed.

wesketchum commented on sbncode/Calibration/TrackCaloSkimmerSelectStoppingTrack_tool.cc:
```diff
@@ -0,0 +1,183 @@
+
+  bool valid_med_dqdx = ((med_dqdx > 0.) && (med_dqdx > fEndMediandQdxCut)) || (fEndMediandQdxCut < 0.);
+
+  std::cout << "Track MED dqdx: " << med_dqdx << " is valid: " << valid_med_dqdx << std::endl;
```
Wrap in a verbosity option.

gputnam replied:
Fixed.

wesketchum commented on sbncode/Calibration/TrackCaloSkimmerSelectStoppingTrack_tool.cc:
```diff
@@ -0,0 +1,183 @@
+
+  bool valid_stopping_residuals = (reduced_residual_offset < fFitResidualsCut) || (fFitResidualsCut < 0.) || !fRequireFit;
+
+  std::cout << "Fit is valid: " << valid_stopping_fit << " " << valid_stopping_expR << " " << valid_stopping_residuals << std::endl;
```
Wrap in a verbosity option.

gputnam replied:
Fixed.

wesketchum commented on sbncode/Calibration/TrackCaloSkimmer_module.cc:
```diff
@@ -0,0 +1,460 @@
+#include "art/Utilities/make_tool.h"
+
+// Global functions / data for fitting
+const size_t MAX_N_FIT_DATA = 30;
```
Make configurable?

gputnam replied:
I don't think I need to -- one shouldn't be fitting more than the last ~5cm of the track and so there's no case where you would need more than 30 points. Plus it's easier to make static with a fixed size.

wesketchum replied:
OK, a little worried that the residual range fit is not configurable ... so I'd recommend at the very least making sure to throw an error if this 30 number is not enough? You may already have that in place, so sorry if you do and I missed it.

wesketchum commented on sbncode/Calibration/TrackCaloSkimmer_module.cc:
```diff
@@ -0,0 +1,460 @@
+const size_t MAX_N_FIT_DATA = 30;
+
+static int N_FIT_DATA = -1;
+static double FIT_RR[MAX_N_FIT_DATA];
```
Use a std::vector? Does this need to be static I guess?

gputnam replied:
When I looked around online this was the suggested way to work with TFitter. I could switch this out for a singleton class in principle, but the fit being performed is simple enough that this works for me.

wesketchum replied:
OK, no strong preference from me here then.

wesketchum commented on sbncode/Calibration/TrackCaloSkimmer_module.cc:
```diff
@@ -0,0 +1,460 @@
+  catch(...) {
+      std::cout << "PFP's with tag: " << fPFPproducer << " not present.\n";
+      // Real data may have missing products -- just ignore the event
+      if (e.isRealData()) return;
```
Do we really want this to fail silently if there is no data? I think we probably do want it to crash in that case...right?

gputnam replied:
For the way ICARUS data has been processed, there are art::Events that have missing data products (if the reconstruction doesn't get run). So this is required functionality for processing such files.

wesketchum replied:
OK, I think this needs some followup with @SFBayLaser  and ICARUS team on cases where reco doesn't get run/missing data products. I've made little conditions/things like this in the past and it's always come back to bite me, as it inevitably hides problems.

If there's a valid reason that we won't have these objects on some events, ok. But reco failing probably means the whole job should fail and we should never get to this point.

gputnam replied:
I don't believe it means the reco is failing. From what I can tell, there is an optical filter that runs and the reconstruction only gets run if the optical filter passes. If the filter fails, then the raw data is still there but the reconstruction does not get run.

cjbackhouse replied:
art filters are a total unintuitive pain. Way out of scope here, but I kind of think the reco should be putting an explicit "failed optical filter" flag in the event, and the downstream recos should be seeing that and, say, putting an empty list of tracks. Then downstream can distinguish between mysteriously missing products and the actual expected program flow.

wesketchum replied:
Not sure I agree @cjbackhouse , but that's a good and different conversation.

@gputnam if that's the case, then the FCL should be configured such that the analysis module only runs on events that pass that filter, using the `SelectEvents: ["name_of_trigger_path"]` in the analysis module configuration (like we do for output streams). I would still remove this more silent failure here.

gputnam replied:
Isn't that only possible to do if the analysis job is in the same larsoft job as the filtering? What if I run on the output of stage0/stage1 processing with the analysis module?

wesketchum commented on sbncode/Calibration/TrackCaloSkimmer_module.cc:
```diff
@@ -0,0 +1,460 @@
+      continue;
+    }
+
+    std::cout << "Processing new track! ID: " << trkPtr->ID() << " time: " << t0 << std::endl;
```
Wrap in verbosity option.

gputnam replied:
Fixed.

wesketchum commented on sbncode/Calibration/TrackCaloSkimmer_module.cc:
```diff
@@ -0,0 +1,460 @@
+  double min = -1;
+
+  for (const sbn::HitInfo &h: hits) {
+    bool hit_is_TPCE = h.tpc <= 1;
```
<=1 ? So, either 0 or 1? Is that right?

gputnam replied:
Yeah. I'd be happy to switch this out for a call to the Geometry service if such a functionality exists.

wesketchum replied:
(see below comment on adding a comment to help document)

wesketchum commented on sbncode/Calibration/TrackCaloSkimmer_module.cc:
```diff
@@ -0,0 +1,460 @@
+    } 
+  }
+
+  return min;
```
This always returns -1 if hit_is_TPCE is false. Is that the desired behavior?

gputnam replied:
Yeah -- the times are non-negative, so -1 is a sensible value to signify no such time exists.

gputnam replied:
On second glance -- I think I misunderstood the comment.

> This always returns -1 if hit_is_TPCE is false. 

This is not correct. If hit_is_TPCE is false, it only accounts for hits in the "West" TPC

wesketchum replied:
OK ,I guess I just don't understand what this function is supposed to be doing then? Can you just briefly walk through that.

gputnam replied:
This function returns the minimum hit time of all the hits on a track in one of the geographic TPC's -- either East or West.

wesketchum replied:
OK, I got it. Just some documentation suggestions then: add a comment that 0 or 1 is 'TPC E' in ICARUS (note, may not be true for SBND...?) and that the function is using the bool TPCE input to say whether you want the east (true) or west (false) TPCs.

I'd actually strongly consider a 'HitMinTimeTPCEast' and 'HitMinTimeTPCWest' function that can call this one underneath just to avoid confusion, as it's really not intuitive.

gputnam replied:
Added comments to document this behavior.

wesketchum commented on sbncode/Calibration/TrackCaloSkimmer_module.cc:
```diff
@@ -0,0 +1,460 @@
+    } 
+  }
+
+  return max;
```
See comments on HitMinTime too.

gputnam replied:
same reply here

wesketchum commented on sbncode/Calibration/TrackCaloSkimmer_module.cc:
```diff
@@ -0,0 +1,460 @@
+  }
+
+  // Save?
+  bool select = false;
```
could alternatively default to true, and only not select if all of the selection tools return false.

gputnam replied:
That's what the code currently does?

wesketchum replied:
Yeah, but it took me a bit to convince myself of that. Modifying the code may make that more clear.

gputnam replied:
It's just De Morgan (https://en.wikipedia.org/wiki/De_Morgan%27s_laws). Also -- one thing I want to do in the piece of code is to save which tool selected the track (so as to tell a cathode crosser from a stopping track, for example). That is most easily done this way.

wesketchum replied:
OK, that's fine. Note if the goal is to know which tool passed, you stop as soon as the first tool passes, so ... you may want to consider a different way to do that.

wesketchum commented on sbncode/Calibration/TrackCaloSkimmer_module.cc:
```diff
@@ -0,0 +1,460 @@
+
+void sbn::TrackCaloSkimmer::DoTailFit() {
+  // Try fitting the constant and exponentials to the tail of dQ/dx v. RR on the collection plane
+  std::vector<double> fit_rr;
```
not sure how big these are, but it may be better to reserve for the total number of hits, so that the push_backs are a bit faster.

gputnam replied:
It shouldn't be more than 10-15 points.

wesketchum replied:
OK. Given it may be doing this frequently, I'd suggest that maybe this be part of the class and just cleared after use ... to avoid allocations in loops and the like. But, if performance is good and it's not worth it, it's fine.

gputnam replied:
I think the performance overhead of this step is minimal.

wesketchum commented on sbncode/Calibration/TrackCaloSkimmer_module.cc:
```diff
@@ -0,0 +1,460 @@
+    FIT_RR[i] = fit_rr[i];
+    FIT_DQDX[i] = fit_dqdx[i];
+
+    std::cout << "FIT: " << FIT_RR[i] << " " << FIT_DQDX[i] << std::endl;
```
Wrap in verbosity option.

gputnam replied:
Fixed.

wesketchum commented on sbncode/Calibration/TrackCaloSkimmer_module.cc:
```diff
@@ -0,0 +1,460 @@
+    fTrack->const_fit_C = C;    
+    fTrack->const_fit_residuals = cresiduals;
+
+    std::cout << "EXP FIT. A: " << A << " R: " << R << " resid: " << residuals << std::endl;
```
Wrap in verbosity option.

gputnam replied:
Fixed.

wesketchum reviewed (changes requested):
Larger suggestions:
* not clear if the tool is very ICARUS focused or not, but if it is could name it "ICARUS" somehow to not worry as much about use in SBND.
* I'm a little worried that the HitInfo object use may just bloat memory. Do you have any memory usage stats for this module from the memory tracker?
* Has this been checked on ICARUS data, with both west and east cryostats? If not, strongly suggest that is done before this is merged in.

gputnam commented:
> Larger suggestions:
> 
>     * not clear if the tool is very ICARUS focused or not, but if it is could name it "ICARUS" somehow to not worry as much about use in SBND.
> 
>     * I'm a little worried that the HitInfo object use may just bloat memory. Do you have any memory usage stats for this module from the memory tracker?
> 
>     * Has this been checked on ICARUS data, with both west and east cryostats? If not, strongly suggest that is done before this is merged in.

I've addressed the first round of comments! On these points -- I thought we wanted it usable in SBND? In any case, I've added a comment to specify its origin in ICARUS. 

The memory footprint of this module is not significant compared to the stage0/stage1 processing. Just did a test which returned <2GB of memory. 

And yes, it has been run in both cryostats on a large number of ICARUS data events.

wesketchum commented:
Thanks Gray! Resolved most, added a few comments. On the bigger ones, that all sounds good, just wanted to be sure.

FNALbuild commented:
CI build for LArSoft on slf7 for e20:prof is in progress -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/11682&builds=)

FNALbuild commented:
CI build for LArSoft on slf7 for c7:prof is in progress -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/11683&builds=)

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/11683&builds=)

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/11682&builds=)

FNALbuild commented:
CI build for SBND on slf7 for c7:prof is in progress -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/6576&builds=)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11683&builds=)

FNALbuild commented:
CI build for ICARUS on slf7 for c7:prof is in progress -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/2044&builds=)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11683&builds=)

FNALbuild commented:
CI build for SBND on slf7 for e20:prof is in progress -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/6577&builds=)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11682&builds=)

FNALbuild commented:
CI build for ICARUS on slf7 for e20:prof is in progress -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/2045&builds=)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11682&builds=)

FNALbuild commented:
:x: CI build for ICARUS Failed at phase build ICARUS on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/2044&builds=)

:rotating_light: For more details about the failed phase, check the [build ICARUS phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/build_detail/phase_details?build_id=icarus_ci/2044&platform=Linux%203.10.0-1160.31.1.el7.x86_64&phase=build&buildtype=slf7%20c7:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11683&builds=)

FNALbuild commented:
:x: CI build for ICARUS Failed at phase build ICARUS on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/2045&builds=)

:rotating_light: For more details about the failed phase, check the [build ICARUS phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/build_detail/phase_details?build_id=icarus_ci/2045&platform=Linux%203.10.0-1160.31.1.el7.x86_64&phase=build&buildtype=slf7%20e20:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11682&builds=)

FNALbuild commented:
:warning: CI build for SBND Warning at phase ci_tests SBND on slf7 for c7:prof - ignored warnings for build -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/6576&builds=)

:rotating_light: For more details about the warning phase, check the [ci_tests SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/6576&platform=Linux%203.10.0-1160.31.1.el7.x86_64&phase=ci_tests&buildtype=slf7%20c7:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11683&builds=)

FNALbuild commented:
:warning: CI build for SBND Warning at phase ci_tests SBND on slf7 for e20:prof - ignored warnings for build -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/6577&builds=)

:rotating_light: For more details about the warning phase, check the [ci_tests SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/6577&platform=Linux%203.10.0-1160.31.1.el7.x86_64&phase=ci_tests&buildtype=slf7%20e20:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11682&builds=)
